pkg/agent/controller/networkpolicy/audit_logging_test.go

```diff
@@ -65,57 +72,81 @@ func newLogInfo(disposition string) (*logInfo, string) {
 	}, expected
 }
 
-func sendMultiplePackets(antreaLogger *AntreaPolicyLogger, ob *logInfo, numPackets int) {
+func sendMultiplePackets(antreaLogger *AntreaPolicyLogger, ob *logInfo, numPackets int, sendInterval time.Duration) {
 	count := 0
-	for range time.Tick(12 * time.Millisecond) {
+	for range time.Tick(sendInterval) {
 		count += 1
-		antreaLogger.logDedupPacket(ob, 100*time.Millisecond)
+		antreaLogger.LogDedupPacket(ob)
 		if count == numPackets {
 			break
 		}
 	}
 }
 
+func closePacketTransmit(mockAnpLogger *mockLogger, testWaitTime time.Duration) {
+	time.Sleep(testWaitTime)
+	mockAnpLogger.mu.Lock()
+	defer mockAnpLogger.mu.Unlock()
```

Comment:
I don't think we need `mockAnpLogger.mu` at all: channels are thread-safe

Reply:
I tried without mutex, it gave this error
`WARNING: DATA RACE
Write at 0x00c00008a730 by goroutine 77:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:357 +0x0
  antrea.io/antrea/pkg/agent/controller/networkpolicy.TestDropPacketMultiDedupLog.func1()
      /usr/src/antrea.io/antrea/pkg/agent/controller/networkpolicy/audit_logging_test.go:127 +0x5e

Previous read at 0x00c00008a730 by goroutine 76:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:158 +0x0
  antrea.io/antrea/pkg/agent/controller/networkpolicy.(*mockLogger).Write()
      /usr/src/antrea.io/antrea/pkg/agent/controller/networkpolicy/audit_logging_test.go:47 +0x1c6
  log.(*Logger).Output()
      /usr/local/go/src/log/log.go:181 +0x3ac
  log.(*Logger).Printf()
      /usr/local/go/src/log/log.go:188 +0x8f
  antrea.io/antrea/pkg/agent/controller/networkpolicy.(*AntreaPolicyLogger).terminateLogKey()
      /usr/src/antrea.io/antrea/pkg/agent/controller/networkpolicy/audit_logging.go:90 +0x191
  antrea.io/antrea/pkg/agent/controller/networkpolicy.(*AntreaPolicyLogger).logAfterTimer()
      /usr/src/antrea.io/antrea/pkg/agent/controller/networkpolicy/audit_logging.go:81 +0xc4`

I think it is because I used mutex when sending to channel to protect the map, so we need a lock here to avoid race condition.

Reply:
actually this is because the race detector believes there is no ordering guarantee between the channel sends and the channel close operation: https://github.com/golang/go/issues/27769. Or believes that a send and the close can happen concurrently. The mutex helps with the second one, so I guess since the error goes away that's what the problem was. There's probably a way to write the code differently so that it's correct even without using a mutex, but I believe the current version is good enough for test code. I'll let @tnqn take a look but this looks good to me.